### ISSUE TITLE
Tracking on Rendering Progress

### ISSUE THREAD

**Issue by benghan95** [inlife/nexrender#57]
Thank you so much for making this library open source! It's super useful for personalised video rendering and I can also generate image thumbnail from your library. 

However, I would like to ask if is there any functions / callback that I can use to keep track on the progress of rendering? For example, I would like to know how many percent left for the rendering process to be completed once I run renderer.render . Or do you have any suggestions on this?

Btw. I'm very new to After Effects CC and aerender. Hope you understand.

**Comment by inlife** [comment]
Hello, thank you for your feedback, really appreciate it !!
Regarding the progress callback, the only solution i could come up with
is just parsing the logs as they appear from the aerender process

for example, you could add console.log there:
https://github.com/inlife/nexrender/blob/master/renderer/tasks/render.js#L98

at least for the purpose of understanding how does it behave
and maybe you will come up with much better way to do this

**Comment by CobraMachete** [comment]
Here is a bit of a brute force method.  Im sure there is a more elegant way of writing this but its still effective.  I modified Vladyslav's highlighted function to look like this:

`    ae.stdout.on('data', (data) => {

        aedata.push(data.toString());
        var str_len = data.toString().split(/[()]+/).filter(Number).length;
        var str_filter = data.toString().split(/[()]+/).filter(Number);

        if (str_len === 1) {
            var percent_complete = ((str_filter/project.settings.endFrame) * 100).toFixed(2);
            console.log(`${percent_complete}% complete`);
        }
    });`

**Comment by craftedbychris** [comment]
Hi, thanks @inlife for making this amazing library, it's extremely useful and I really appreciate all your hard work. Thanks @CobraMachete for that piece of code, it  just what I had in mind. I was just wondering if anyone knew the best way to pass the percent_complete var back (or the data stream) to this part of the API setup in my original node app. 

`project.on('rendering', function(err, project) {
 });`

Tried everything I can think of. Any help would be amazing :)

**Comment by inlife** [comment]
Thank you @craftedbychris for your feedback, very pleased to hear that :)

As far as I understand your use case is node app, with API connection to master ?
In that case it kinda hard to get such a constant feed of information back to the client from render node via HTTP. This is possible, however it requires a lot of hustle around. 

In case of @CobraMachete that was a local client that he just printed info from via console.log. So that was much easier.

Maybe I didnt understand you properly tho, and you have slightly different use case?

Thank you.

**Comment by craftedbychris** [comment]
Hi @inlife, thanks for the quick reply I appreciate it. Yes that;s right, I have a main node app calling API to render, I modified the code from the - Usage (Render network + node.js API wrapper) example on the homepage. 

I'm then running nexrender api-server & nexrender renderer, both on port 8000. Everything is working great, getting videos back as they should be. @CobraMachete's code works but the console log prints to the nexrender window.

Any ideas in the best direction to head to get it back to the node app itself? Would it be best to read the ae log file into the main app and process it there?

Many Thanks :)

**Comment by craftedbychris** [comment]
Yes that sounds like the easiest and most direct option I think. I will play around today and let you know what happens.

Thanks for your time @inlife :)

**Comment by benghan95** [comment]
Thank you everyone for the methods given! Much appreciate. 🥁 
I guess I will try out the data streaming on log file one then. 
Thanks again for this awesome library. :)

**Comment by inlife** [comment]
hey, thats a job anyway, even tho that sounds funny, still great :D

**Comment by inlife** [comment]
worker-side rendering progress has been added in the latest pre-release version, closing the issue
